### PR TITLE
Use a dedicated `PrimRef` for `TypeRef`s of primitive types.

### PR DESCRIPTION
**Pull request by sjrd**
Before, we put primitive types together with class types as `ClassRef`s, using magical "class names" that actually represented primitives. Those special `ClassRef`s for primitives had to be special-cased virtually everywhere, by testing whether the class name was one of the magical names.

Now, we have a dedicated `PrimRef` instead. This simplifies a lot of code paths by using pattern-matching instead of tests for the class name. More interestingly, it allows to completely get rid of the magical class names.

The char codes associated with primitive type refs are now only used by the method signature encodings, and by some emitter-specific code (and although they use the same characters, that is simply by convention; it is not required for correctness).

### CODE REVIEW

File: compiler/src/main/scala/org/scalajs/nscplugin/TypeConversions.scala [hunk 1]
```diff
@@ -40,19 +40,19 @@ trait TypeConversions[G <: Global with Singleton] extends SubComponent {
...
+        BooleanClass -> Types.BooleanRef,
+        CharClass    -> Types.CharRef,
+        ByteClass    -> Types.ByteRef,
+        ShortClass   -> Types.ShortRef,
+        IntClass     -> Types.IntRef,
+        LongClass    -> Types.LongRef,
+        FloatClass   -> Types.FloatRef,
+        DoubleClass  -> Types.DoubleRef,
+        NothingClass -> Types.ClassRef(encodeClassFullName(RuntimeNothingClass)),
+        NullClass    -> Types.ClassRef(encodeClassFullName(RuntimeNullClass))
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/nscplugin/TypeConversions.scala]: I know this is already here, but I don't understand why this is not simply the primitive type ref here...

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 2]
```diff
@@ -189,21 +163,53 @@ object Definitions {
...
+      case ArrayTypeRef(base, dimensions) =>
+        val encodedBase = encodeTypeRef(base)
+        val builder = new java.lang.StringBuilder(dimensions + encodedBase.length())
+        var i = 0
+        while (i != dimensions) {
+          builder.append('A')
+          i += 1
+        }
+        builder.append(encodedBase)
+        builder.toString()
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Why not simply `"A" * dimensions + encodeTypeRef(base)`?

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 3]
```diff
@@ -189,21 +163,53 @@ object Definitions {
...
     }
   }
 
   /** Decodes a [[Types.TypeRef]], such as in an encoded method signature.
    */
   def decodeTypeRef(encodedName: String): TypeRef = {
-    val arrayDepth = encodedName.indexWhere(_ != 'A')
+    var arrayDepth = 0
+    while (encodedName.charAt(arrayDepth) == 'A')
+      arrayDepth += 1
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Why is this changed?

File: ir/src/main/scala/org/scalajs/ir/Types.scala [hunk 4]
```diff
@@ -163,17 +163,41 @@ object Types {
...
+  final val BooleanRef = PrimRef(BooleanType)("Z", Tags.TagBooleanRef)
+  final val CharRef = PrimRef(CharType)("C", Tags.TagCharRef)
+  final val ByteRef = PrimRef(ByteType)("B", Tags.TagByteRef)
+  final val ShortRef = PrimRef(ShortType)("S", Tags.TagShortRef)
+  final val IntRef = PrimRef(IntType)("I", Tags.TagIntRef)
+  final val LongRef = PrimRef(LongType)("J", Tags.TagLongRef)
+  final val FloatRef = PrimRef(FloatType)("F", Tags.TagFloatRef)
+  final val DoubleRef = PrimRef(DoubleType)("D", Tags.TagDoubleRef)
+  final val NullRef = PrimRef(NullType)("N", Tags.TagNullRef)
+  final val NothingRef = PrimRef(NothingType)("E", Tags.TagNothingRef)
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Types.scala]: I think it would make sense to provide an accessor per `encodedName` here, just like in an enum. The split of this logic seems a bit strange.

The same applies to tags.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala [hunk 5]
```diff
@@ -310,31 +312,47 @@ object Infos {
...
-      addAccessedClassData(baseNameOf(tpe))
+    def maybeAddAccessedClassData(typeRef: TypeRef): this.type = {
+      typeRef match {
+        case ClassRef(cls) =>
+          addAccessedClassData(cls)
+        case ArrayTypeRef(ClassRef(cls), _) =>
+          addAccessedClassData(cls)
+        case _ =>
+      }
+      this
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala]: Create a helper for this pattern match? It appears many times.

Something like:

```scala
private def maybeAdd(typeRef: TypeRef, add: String => this.type): this.type = {
 // ...
}
```

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala [hunk 6]
```diff
@@ -350,15 +352,37 @@ private[emitter] final class JSGen(val semantics: Semantics,
...
+    case CharRef    => "C"
+    case ByteRef    => "B"
+    case ShortRef   => "S"
+    case IntRef     => "I"
+    case LongRef    => "J"
+    case FloatRef   => "F"
+    case DoubleRef  => "D"
+    case NullRef    => "N"
+    case NothingRef => "E"
+  }
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala]: I find it a bit strange that we introduce types but the emitter internally still uses char codes in many locations (n.b. the corejslib). Consider using this function everywhere to translate.

File: linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala [hunk 7]
```diff
@@ -1199,29 +1199,16 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
...
-        case 'B' => ByteType
-        case 'S' => ShortType
-        case 'I' => IntType
-        case 'J' => LongType
-        case 'F' => FloatType
-        case 'D' => DoubleType
-        case 'N' => NullType
-        case 'E' => NothingType
-        case 'O' => AnyType
-        case 'T' => ClassType(BoxedStringClass) // NOT StringType
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala]: Huh... What happened here? (Was this forgotten in the previous PR)?

File: linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala [hunk 8]
```diff
@@ -1236,9 +1223,15 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
...
       implicit ctx: ErrorContext): Type = {
-    val ArrayTypeRef(baseClassName, dimensions) = arrayTypeRef
-    if (dimensions == 1) classNameToType(baseClassName)
-    else ArrayType(ArrayTypeRef(baseClassName, dimensions - 1))
+    val ArrayTypeRef(base, dimensions) = arrayTypeRef
+    if (dimensions == 1) {
+      base match {
+        case PrimRef(tpe)  => tpe
+        case ClassRef(cls) => classNameToType(cls)
+      }
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala]: Simply call `classRefToType`?

File: linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala [hunk 9]
```diff
@@ -71,7 +71,7 @@ object TestIRBuilder {
   }
 
   def mainMethodDef(body: Tree): MethodDef = {
-    val stringArrayType = ArrayType(ArrayTypeRef("T", 1))
+    val stringArrayType = ArrayType(ArrayTypeRef(ClassRef(BoxedStringClass), 1))
```
Comment (gzm0) [inline linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala]: I also do not understand this :)

Comment (sjrd) [reply linker/shared/src/test/scala/org/scalajs/linker/testutils/TestIRBuilder.scala]: Hum?

1. `"T"` -> `ClassRef("T")` to adapt to the new arguments of `ArrayTypeRef`
2. Replace the literal `"T"` by the named constant `BoxedStringClass` which is equal to `"T"`

Perhaps you thought `"T"` was for *primitive* strings rather than "boxed" strings? It's a historical anomaly that `"T"` stands for `java.lang.String` rather than the primitive `StringType`. It was like that for shortening encoded names as much as possible. This thing will disappear once I get to delaying mangling to the Emitter, which should hopefully come after this PR (this PR came to fix difficulties I encountered when trying to implement the delayed mangling, actually).

File: ir/src/main/scala/org/scalajs/ir/Types.scala [hunk 10]
```diff
@@ -163,17 +163,38 @@ object Types {
...
+  final val CharRef = PrimRef(CharType)
+  final val ByteRef = PrimRef(ByteType)
+  final val ShortRef = PrimRef(ShortType)
+  final val IntRef = PrimRef(IntType)
+  final val LongRef = PrimRef(LongType)
+  final val FloatRef = PrimRef(FloatType)
+  final val DoubleRef = PrimRef(DoubleType)
+  final val NullRef = PrimRef(NullType)
+  final val NothingRef = PrimRef(NothingType)
+
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Types.scala]: Actually: Any reason these are not case objects? We could avoid the unchecked pattern matches then, no? Also it would be more consistent with the primitive types.

File: ir/src/main/scala/org/scalajs/ir/Serializers.scala [hunk 11]
```diff
@@ -673,6 +673,20 @@ object Serializers {
     }
 
     def writeTypeRef(typeRef: TypeRef): Unit = typeRef match {
+      case typeRef: PrimRef =>
+        (typeRef: @unchecked) match {
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Serializers.scala]: For example here, case objects would help, no?

File: ir/src/main/scala/org/scalajs/ir/Types.scala (hunk 10 above)
Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Types.scala]: Yeah, I thought a lot about that. I really wanted to make them case objects, but if I do that I cannot make `PrimRef` a case class, and it really should be a case class for pattern matching of the form `case PrimRef(tpe)` to be efficient. So I didn't really have a choice, unfortunately.

We could do an abominable thing and make `PrimRef` and `abstract case class`, while making the individuals (non-case) `object`s ... But if I do that I won't be able to argue with Martin that abstract case class are an abomination anymore.

Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Types.scala]: Right. And if you simply define an extractor, you do not get exhaustiveness checking.

But, now I can of course come back to a slight variation of `PrimType`: A sub hierarchy (`RefablePrimType`) would allow exhaustiveness checking on the inner type. The `final val`s could be kept as mere aliases or not.

In the same swoop, I'd avoid overriding `equals` and `hashCode`: It feels like premature optimization traded off for some safety.

WDYT?

Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Types.scala]: Yes, that's probably reasonable. I'll try that later today when I get the chance.

Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Types.scala]: Done.
